src/datatable/utils/fread.py

```diff
@@ -238,21 +248,47 @@ def _resolve_archive(filename, subpath, tempfiles):
                 filename += "/" + subpath
                 zff = [subpath]
             else:
-                raise ValueError("File `%s` does not exist in archive `%s`"
-                                 % (subpath, filename))
-        if len(zff) > 1:
-            warnings.warn("Zip file %s contains multiple compressed "
-                          "files: %r. Only the first of them will be used."
-                          % (filename, zff), category=FreadWarning)
-            filename += "/" + zff[0]
-        if len(zff) == 0:
-            raise ValueError("Zip file %s is empty" % filename)
-        if logger:
-            logger.debug("Extracting %s to temporary directory %s"
-                              % (filename, tempfiles.tempdir))
-        newfile = zf.extract(zff[0], path=tempfiles.tempdir)
-        tempfiles.add(newfile)
-        out_file = newfile
+                raise IOError("File `%s` does not exist in archive `%s`"
+                               % (subpath, filename))
+        extracted_files = []
+        for zf_file in zff:
+            if logger:
+                logger.debug("Extracting %s/%s to temporary directory %s"
+                             % (filename, zf_file, tempfiles.tempdir))
+            newfile = zf.extract(zf_file, path=tempfiles.tempdir)
+            srcname = filename + "/" + zf_file
+            tempfiles.add(newfile)
+            extracted_files.append(((srcname, newfile, None, None), None))
+        if len(extracted_files) == 1:
+            out_file = extracted_files[0][0][1]
+        else:
+            return (None, None, None, None), extracted_files
+
+    elif filename.endswith(".tar.gz"):
+        import tarfile
+        zf = tarfile.open(filename, mode="r:gz")
+        zff = [entry.name for entry in zf.getmembers() if entry.isfile()]
+        if subpath:
+            if subpath in zff:
+                filename += "/" + subpath
```

Comment:
`os.path.join()` here too?

Reply:
Yeah, I've thought about it, but not sure about the correct approach.
Because  these are "fake" paths, where the name of an archive is treated as a folder: `data.zip/train.csv`. Is it better to use forward or backslash here? Do we need to perform slash-normalization? How does `zff` array look like, esp. if the archive has some nested directory structure inside?

Reply:
Good questions. I remember sometimes depending on the source OS an archive could have different slashes :/ So this needs to be investigated in more details.